### PR TITLE
Show a readable name in mz_cluster_replicas.size

### PR DESCRIPTION
**Pull request by teskje**
Before:

```sql
materialize=> select * from mz_cluster_replicas;
 cluster_id | id |      name       | size | availability_zone | status
------------+----+-----------------+------+-------------------+---------
          2 |  2 | r1              | 2-1  |                   | healthy
          1 |  1 | default_replica | 1-1  |                   | healthy
```

After:

```sql
materialize=> select * from mz_cluster_replicas;
 cluster_id | id |      name       |  size   | availability_zone | status
------------+----+-----------------+---------+-------------------+---------
          2 |  2 | r1              | small   |                   | healthy
          1 |  1 | default_replica | default |                   | healthy
```

### Motivation

  * This PR fixes a previously unreported bug.

    `mz_cluster_replicas.size` contains the string `"{size_config.scale}-{size_config.workers}"`, not the size name the user specifies when creating a replica. Without any further explanation for what the string means, users will find this confusing. Plus, we might want to hide these sizing details from users anyway.

    While trying to write a test, I discovered a second bug that is also fixed here:

    ```sql
    materialize=> CREATE CLUSTER REPLICA default."test-123" SIZE '1';
    CREATE CLUSTER REPLICA
    materialize=> DROP CLUSTER REPLICA default."test-123";
    ERROR:  CLUSTER default has no CLUSTER REPLICA named "test-123"
    ```

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Change the `size` field of the `mz_cluster_replicas` table to contain a readable size name.

### CODE REVIEW

File: src/coord/src/catalog/storage.rs [hunk 1]
```diff
@@ -150,19 +150,25 @@ async fn migrate<S: Append>(stash: &mut S, version: u64) -> Result<(), StashErro
...
                                 name: "default_replica".into(),
-                                config: "{\"Managed\":{\"size_config\":{\"memory_limit\": null, \"cpu_limit\": null, \"scale\": 1, \"workers\": 1}}}".into(),
-                            }
-                        )
-                    ]
-                ).await?;
+                                config: "{ \
+                                    \"Managed\": { \
+                                        \"size_config\": {\"scale\": 1, \"workers\": 1}, \
+                                        \"size_name\": \"default\" \
```
Comment (teskje) [inline src/coord/src/catalog/storage.rs:164]: We don't have access to the size config map here, to find the user-facing name for this size config. But I'm not sure if "default" is a helpful name. Maybe we should just set this to the empty string?

File: src/dataflow-types/src/client/controller.rs [hunk 2]
```diff
@@ -328,8 +329,9 @@ where
         config: ConcreteComputeInstanceReplicaConfig,
     ) -> Result<(), anyhow::Error> {
         if let ConcreteComputeInstanceReplicaConfig::Managed {
-            size_config: _size_config,
-            availability_zone: _az,
+            size_config: _,
+            size_name: _,
+            availability_zone: _,
```
Comment (andrioni) [inline src/dataflow-types/src/client/controller.rs:334]: Weak suggestion: Maybe `if let ConcreteComputeInstanceReplicaConfig::Managed { .. }` would be cleaner?

**Review by andrioni — APPROVED** [review]
LGTM, could you maybe add a TD/SLT test with the new behavior?

File: src/dataflow-types/src/client/controller.rs (hunk 2 above)
Comment (teskje) [reply src/dataflow-types/src/client/controller.rs:334]: @sploiselle's opinion is that we should list all fields here so developers are nudged into thinking about new fields affecting the below code (https://github.com/MaterializeInc/materialize/pull/12313#discussion_r869310618). I think this is reasonable, but I'd be fine either way.

**Review by umanwizard — APPROVED** [review]

**Comment by sploiselle** [comment]
I think this change should be reverted, or an additional PR needs to be made to let this table's output work as input:

```
materialize=> CREATE CLUSTER foo REPLICAS (r1 (size 'default'));
ERROR:  unknown cluster replica size default
HINT:  Valid cluster replica sizes are: 8, 2, 32, 16, 1, 4, 2-1, 2-2, 2-4

materialize=> CREATE CLUSTER foo REPLICAS (r1 (size 'small'));
ERROR:  unknown cluster replica size small
HINT:  Valid cluster replica sizes are: 8, 2, 32, 16, 1, 4, 2-1, 2-2, 2-4
```

The latter is preferable but requires working with the cloud team's descriptions of what the size parameters mean.

**Comment by teskje** [comment]
@sploiselle I'm not sure how reverting the PR would make anything better in this regard? This is the situation after this PR:

* `bin/materialized`:

```sql
materialize=> SELECT * FROM mz_cluster_replicas;
 cluster_id | id |      name       |  size   | availability_zone | status
------------+----+-----------------+---------+-------------------+---------
          2 |  2 | r1              | 2-1     |                   | unknown
          1 |  1 | default_replica | default |                   | unknown

materialize=> CREATE CLUSTER foo2 REPLICAS (r1 (size '2-1'));
CREATE CLUSTER
materialize=> CREATE CLUSTER foo2 REPLICAS (r1 (size 'default'));
ERROR:  unknown cluster replica size default
HINT:  Valid cluster replica sizes are: 16, 8, 2, 1, 32, 4, 2-2, 2-1, 2-4
```

* Cloud:

```sql
materialize=> SELECT * FROM mz_cluster_replicas;
 cluster_id | id |      name       |  size   | availability_zone | status
------------+----+-----------------+---------+-------------------+---------
          2 |  2 | r1              | small   |                   | healthy
          1 |  1 | default_replica | default |                   | healthy

materialize=> CREATE CLUSTER foo2 REPLICAS (r1 (size 'small'));
CREATE CLUSTER
materialize=> CREATE CLUSTER foo2 REPLICAS (r1 (size 'default'));
ERROR:  unknown cluster replica size default
HINT:  Valid cluster replica sizes are: xsmall, small, medium, large, xlarge, 2xlarge, 3xlarge, 4xlarge, 5xlarge, 6xlarge
```

So the table's output works as input both locally and on cloud, except for the default replica. Whereas prior to this change, the table's output worked as input always locally but never on cloud, which seems worse to me.

I agree that the "default" size situation is not great (see my comment above), but I'm not sure what would be a good solution to this. When we pack the default replica into stash we don't have any information about replica size names in the current environment.

**Comment by teskje** [comment]
I guess we could make it so that the size "default" is always understood as (scale=1, workers=1), irrespectively of the current size configuration.

**Comment by sploiselle** [comment]
@teskje My mistake here--I had thought the mapping introduced here was the numeric format -> t-shirt size format, and I didn't test it on cloud to see that working properly. Apologies, and really appreciate you taking the time to make this fix.

**Comment by teskje** [comment]
Sure, no worries :)

**Pull request merged**